File: src/org/dgraph/graph/algorithm/AStarSearch.java
package org.dgraph.graph.algorithm;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.function.BiFunction;

import org.dgraph.graph.Graph;
import org.dgraph.graph.edge.WeightedEdge;
import org.dgraph.graph.path.SimpleWeightedPath;
import org.dgraph.graph.path.WeightedPath;
import org.dgraph.util.FibonacciHeap;
import org.dgraph.util.Tuple;
import org.dgraph.util.FibonacciHeap.Node;

/** 
 * 
 *
 * @author Andrii Dzhyrma 
 * @since 1.0
 */
public class AStarSearch {

	public static <V, E extends WeightedEdge<V, W>, W> WeightedPath<V, E> findShortestPath(Graph<V, E> graph, V source, V target,
			BiFunction<V, V, Double> heuristic) {

		/* Keep track on previous vertices, to reconstruct the shortest path. */
		HashMap<V, E> previous = new HashMap<>(graph.sizeOfVertices());
		
		/* Keep track on vertices with distance passed inserted to Fibonacci heap. */
		HashMap<V, Node<Tuple<V, Double>>> heapNodes = new HashMap<>(graph.sizeOfVertices());
		FibonacciHeap<Tuple<V, Double>> heap = new FibonacciHeap<>();

		/* Insert the source vertex to the heap. */
		heapNodes.put(source, heap.enqueue(new Tuple<>(source, 0d), 0d));
		while (!heap.isEmpty()) {
			Node<Tuple<V, Double>> cur = heap.dequeueMin();

			/* Target vertex is reached. */
			if (cur.equals(target))
				break;

			/* Go through each edge from the currently picked vertex. */
			for (E e : graph.getEdgesFromSource(cur.getValue().getItem1())) {
				
				/* A* search algorithm works only with non-negative weights. */
				if (e.getWeight() < 0)
					throw new IllegalArgumentException("A* search algorithm can be applied only for graphs with non negative weights.");
				
				V adj = e.getTarget();
				
				/* Calculate the distance using a sum of the passed distance and weight of the edge. */
				double newDistance = cur.getValue().getItem2().doubleValue() + e.getWeight();
				
				Node<Tuple<V, Double>> next = heapNodes.get(adj);
				
				/* If this vertex has never been added to the heap or its previous distance was larger the the new one. */
				if (next == null || next.getValue().getItem2().doubleValue() > newDistance) {
					/* Check, whether the vertex is in the heap. */
					if (next == null || next.isDequeued())
						heapNodes.put(adj, heap.enqueue(new Tuple<>(adj, newDistance), newDistance + heuristic.apply(adj, target)));
					/* Otherwise decrease the key of the node in the heap. */
					else
						heap.decreaseKey(next, newDistance + heuristic.apply(adj, target));
					
					/* Change a preceding vertex. */
					previous.put(adj, e);
				}
			}
		}
		
		/* Check, whether target node has been reached. */
		if (!heapNodes.containsKey(target))
			return null;
		
		/* Path reconstruction. */
		LinkedList<E> edges = new LinkedList<>();
		V cur = target;
		E curEdge;
		while (!cur.equals(source)) {
			curEdge = previous.get(cur);
			edges.push(curEdge);
			cur = curEdge.getSource();
		}
		return new SimpleWeightedPath<>(source, target, edges);
	}
}

File: src/main/kotlin/org/dzhyrma/dgraph/algo/AStarSearch.kt
package org.dzhyrma.dgraph.algo

import org.dzhyrma.dgraph.Graph
import org.dzhyrma.dgraph.edge.WeightedEdge
import org.dzhyrma.dgraph.forEachEdge
import org.dzhyrma.dgraph.path.SimpleWeightedPath
import org.dzhyrma.dgraph.path.WeightedPath
import org.dzhyrma.dgraph.util.FibonacciHeap
import java.util.LinkedList

/**
 * A* is an informed search algorithm, or a best-first search, meaning that it is formulated in terms of weighted
 * graphs: starting from a specific starting node of a graph, it aims to find a path to the given goal node having
 * the smallest cost (least distance travelled, shortest time, etc.). It does this by maintaining a tree of paths
 * originating at the start node and extending those paths one edge at a time until its termination criterion is
 * satisfied.
 *
 * Peter Hart, Nils Nilsson and Bertram Raphael of Stanford Research Institute (now SRI International) first published
 * the algorithm in 1968.[3] It can be seen as an extension of Edsger Dijkstra's 1959 algorithm. A* achieves better
 * performance by using heuristics to guide its search.
 */
object AStarSearch {

    /**
     * Performs an A* search on a weighted graph and finds one of the optimal paths.
     *
     * @param graph weighted graph
     * @param source source vertex, from which the search should start
     * @param target target vertex, where the search should end
     * @param heuristic admissible function that estimates the cost of the cheapest path from a vertex to the target
     * @return a shortest(if [heuristic] is admissible) path if it is found, null otherwise
     */
    fun <V, E : WeightedEdge<V>> perform(
        graph: Graph<V, E>,
        source: V,
        target: V,
        heuristic: (V) -> Double,
    ): WeightedPath<V, E>? {
        val bestPathEdges = HashMap<V, E>(graph.vertices.size)

        val heapNodes = HashMap<V, FibonacciHeap.Node<PathNode<V>>>(graph.vertices.size)
        val heap = FibonacciHeap<PathNode<V>>()

        heapNodes[source] = heap.enqueue(PathNode(source, 0.0), heuristic(source))
        var targetReached = false
        while (!targetReached) {
            val min = heap.dequeueMin() ?: return null
            val (currentVertex, bestDistance) = min.value
            when (currentVertex) {
                target -> targetReached = true
                else -> prioritizeAdjacentVertices(
                    graph = graph,
                    currentVertex = currentVertex,
                    bestDistance = bestDistance,
                    heapNodes = heapNodes,
                    heap = heap,
                    heuristic = heuristic,
                    bestPathEdges = bestPathEdges,
                )
            }
        }

        return reconstructPath(target, source, bestPathEdges)
    }

    private fun <E : WeightedEdge<V>, V> prioritizeAdjacentVertices(
        graph: Graph<V, E>,
        currentVertex: V,
        bestDistance: Double,
        heapNodes: HashMap<V, FibonacciHeap.Node<PathNode<V>>>,
        heap: FibonacciHeap<PathNode<V>>,
        heuristic: (V) -> Double,
        bestPathEdges: HashMap<V, E>,
    ) = graph.forEachEdge(currentVertex) { edge, nextVertex, weight ->
        check(weight >= 0) { "A* search algorithm can be applied only for graphs with non negative weights." }

        val nextHeapNode = heapNodes[nextVertex]
        val distanceToNextVertex = bestDistance + weight
        if (nextHeapNode == null || distanceToNextVertex < nextHeapNode.value.bestDistance) {
            val nextVertexPriority = distanceToNextVertex + heuristic(nextVertex)
            changeNextVertexPriority(
                nextHeapNode = nextHeapNode,
                heapNodes = heapNodes,
                heap = heap,
                nextVertex = nextVertex,
                distanceToNextVertex = distanceToNextVertex,
                nextVertexPriority = nextVertexPriority
            )
            bestPathEdges[nextVertex] = edge
        }
    }

    private fun <V> changeNextVertexPriority(
        nextHeapNode: FibonacciHeap.Node<PathNode<V>>?,
        heapNodes: HashMap<V, FibonacciHeap.Node<PathNode<V>>>,
        heap: FibonacciHeap<PathNode<V>>,
        nextVertex: V,
        distanceToNextVertex: Double,
        nextVertexPriority: Double,
    ) {
        when {
            nextHeapNode == null || nextHeapNode.isDequeued -> {
                heapNodes[nextVertex] = heap.enqueue(
                    value = PathNode(nextVertex, distanceToNextVertex),
                    priority = nextVertexPriority,
                )
            }
            else -> {
                nextHeapNode.value.bestDistance = distanceToNextVertex
                heap.decreaseKey(nextHeapNode, nextVertexPriority)
            }
        }
    }

    private fun <E : WeightedEdge<V>, V> reconstructPath(
        target: V,
        source: V,
        bestPathEdges: HashMap<V, E>,
    ): SimpleWeightedPath<V, E>? {
        val edges = LinkedList<E>()
        var current: V = target
        var curEdge: E
        while (current != source) {
            curEdge = bestPathEdges[current] ?: return null
            edges.push(curEdge)
            current = if (curEdge.source == current) curEdge.target else curEdge.source
        }
        return SimpleWeightedPath(source, target, edges)
    }

    private data class PathNode<V>(
        val vertex: V,
        var bestDistance: Double,
    )
}
